src/v/cluster/partition_balancer_planner.cc

```diff
@@ -358,6 +358,105 @@ void partition_balancer_planner::get_unavailable_nodes_reassignments(
     }
 }
 
+/// Try to fix ntps that have several replicas in one rack (these ntps can
+/// appear because rack awareness constraint is not a hard constraint, e.g. when
+/// a rack dies and we move all replicas that resided on dead nodes to live
+/// ones).
+///
+/// We go over all such ntps (a list maintained by partition_balancer_state) and
+/// if the number of currently live racks is more than the number of racks that
+/// the ntp is replicated to, we try to schedule a move. For each rack we
+/// arbitrarily choose the first appearing replica to remain there (note: this
+/// is probably not optimal choice).
+void partition_balancer_planner::get_rack_constraint_repair_reassignments(
+  plan_data& result, reallocation_request_state& rrs) {
+    if (_state.ntps_with_broken_rack_constraint().empty()) {
+        return;
+    }
+
+    absl::flat_hash_set<model::rack_id> available_racks;
+    for (auto node_id : rrs.all_nodes) {
+        if (!rrs.timed_out_unavailable_nodes.contains(node_id)) {
+            auto rack = _partition_allocator.state().get_rack_id(node_id);
+            if (rack) {
+                available_racks.insert(*rack);
+            }
+        }
+    }
+
+    for (const auto& ntp : _state.ntps_with_broken_rack_constraint()) {
+        if (rrs.planned_moves_size >= _config.movement_disk_size_batch) {
+            return;
+        }
+
+        if (rrs.moving_partitions.contains(ntp)) {
+            continue;
+        }
+
+        auto assignment = _state.topics().get_partition_assignment(ntp);
+        if (!assignment) {
+            vlog(clusterlog.warn, "assignment for ntp {} not found", ntp);
+            continue;
+        }
+
+        const auto& orig_replicas = assignment->replicas;
+
+        std::vector<model::broker_shard> stable_replicas;
+        absl::flat_hash_map<model::rack_id, model::broker_shard>
```

Comment:
it seems that the `flat_hash_set` should be enough as the value is never used

Reply:
nice catch 👍